File: src/cli/commands/board.js
import chalk from 'chalk';
import { findProjectRoot, loadConfig } from '../../lib/config.js';
import { createBoard } from '../../lib/board.js';

function getBoard() {
  const root = findProjectRoot();
  if (!root) {
    console.error(chalk.red('No .holoctl/ found. Run `holoctl init` first.'));
    process.exit(1);
  }
  const config = loadConfig(root);
  return { board: createBoard(root, config), config, root };
}

export function registerBoardCommand(program) {
  const board = program
    .command('board')
    .alias('b')
    .description('Manage the project board (tickets, statuses, sprints)');

  board
    .command('stat')
    .description('Show ticket counts by status')
    .action(() => {
      const { board: b } = getBoard();
      const stats = b.stat();
      console.log(JSON.stringify(stats, null, 2));
    });

  board
    .command('get <id>')
    .description('Get a single ticket by ID')
    .action((id) => {
      const { board: b } = getBoard();
      const ticket = b.get(id);
      if (!ticket) {
        console.error(chalk.red(`Ticket ${id} not found`));
        process.exit(1);
      }
      console.log(JSON.stringify(ticket, null, 2));
    });

  board
    .command('ls')
    .description('List tickets with optional filters')
    .option('--sprint <sprint>', 'Filter by sprint')
    .option('--status <status>', 'Filter by status')
    .option('--agent <agent>', 'Filter by agent')
    .option('--tag <tag>', 'Filter by tag')
    .option('--project <name>', 'Filter by project (subdir name discovered in workspace)')
    .argument('[priority]', 'Filter by priority (p0, p1, p2, p3)')
    .action((priority, opts) => {
      const { board: b } = getBoard();
      const filters = { ...opts };
      if (priority && /^p\d$/.test(priority)) filters.priority = priority;

      const tickets = b.ls(filters);
      if (tickets.length === 0) {
        console.log(chalk.dim('No tickets match the filters.'));
        return;
      }

      for (const t of tickets) {
        const dep = t.depends?.length ? chalk.dim(` [dep: ${t.depends.join(', ')}]`) : '';
        const agents = t.agent?.length ? chalk.green(t.agent.join(',')) : chalk.dim('—');
        console.log(
          `${chalk.bold(t.id)}  ${priorityColor(t.priority)}  ${statusColor(t.status)}  ${(t.sprint || '—').padEnd(12)}  ${agents.padEnd(20)}  ${t.title.slice(0, 50)}${dep}`
        );
      }
    });

  board
    .command('move <id> <status>')
    .description('Move a ticket to a new status')
    .action((id, status) => {
      const { board: b } = getBoard();
      try {
        const result = b.move(id, status);
        console.log(`${result.id}: ${result.from} → ${chalk.bold(result.to)}`);
      } catch (e) {
        console.error(chalk.red(e.message));
        process.exit(1);
      }
    });

  board
    .command('set <id> <field> [value...]')
    .description('Set a field on a ticket')
    .action((id, field, valueParts) => {
      const { board: b } = getBoard();
      const value = valueParts.join(' ');
      try {
        const result = b.set(id, field, value);
        console.log(`${result.id}.${result.field} = ${JSON.stringify(result.value)}`);
      } catch (e) {
        console.error(chalk.red(e.message));
        process.exit(1);
      }
    });

  board
    .command('add <json>')
    .description('Create a new ticket from JSON')
    .action((json) => {
      const { board: b } = getBoard();
      try {
        const patch = JSON.parse(json);
        const ticket = b.add(patch);
        console.log(chalk.green(`Created ${ticket.id}: ${ticket.title}`));
        console.log(JSON.stringify(ticket, null, 2));
      } catch (e) {
        console.error(chalk.red(e.message));
        process.exit(1);
      }
    });

  board
    .command('next-id')
    .description('Show the next available ticket ID')
    .action(() => {
      const { board: b } = getBoard();
      console.log(b.nextId());
    });

  board
    .command('rebuild-index')
    .description('Rebuild index.json from ticket .md files')
    .action(() => {
      const { board: b } = getBoard();
      const result = b.rebuildIndex();
      console.log(chalk.green(`Rebuilt index: ${result.ticketCount} tickets, nextId: ${result.nextId}`));
    });
}

function priorityColor(p) {
  const colors = { p0: chalk.red, p1: chalk.yellow, p2: chalk.blue, p3: chalk.dim };
  return (colors[p] || chalk.white)(p.padEnd(2));
}

function statusColor(s) {
  const colors = { backlog: chalk.dim, doing: chalk.cyan, review: chalk.yellow, done: chalk.green, cancelled: chalk.strikethrough };
  return (colors[s] || chalk.white)(s.padEnd(10));
}

File: projhub/cli/board.py
from __future__ import annotations
import json
from pathlib import Path
from typing import Optional

import typer
from ._console import console

from ..lib.config import find_project_root, load_config
from ..lib.board import Board

app = typer.Typer(help="Manage the project board (tickets, statuses, sprints)")


def _get_board() -> tuple[Board, dict, Path]:
    root = find_project_root()
    if not root:
        console.print("[red]No .projhub/ found. Run `projhub init` first.[/red]")
        raise typer.Exit(1)
    config = load_config(root)
    return Board(root, config), config, root


@app.command("stat")
def stat_cmd():
    """Show ticket counts by status."""
    board, _, _ = _get_board()
    print(json.dumps(board.stat(), indent=2))


@app.command("get")
def get_cmd(ticket_id: str = typer.Argument(..., help="Ticket ID")):
    """Get a single ticket by ID."""
    board, _, _ = _get_board()
    ticket = board.get(ticket_id)
    if not ticket:
        console.print(f"[red]Ticket {ticket_id} not found[/red]")
        raise typer.Exit(1)
    print(json.dumps(ticket, indent=2))


@app.command("ls")
def ls_cmd(
    priority: Optional[str] = typer.Argument(None, help="Filter by priority (p0, p1, p2, p3)"),
    sprint: Optional[str] = typer.Option(None, "--sprint"),
    status: Optional[str] = typer.Option(None, "--status"),
    agent: Optional[str] = typer.Option(None, "--agent"),
    tag: Optional[str] = typer.Option(None, "--tag"),
    scope: Optional[str] = typer.Option(None, "--scope"),
):
    """List tickets with optional filters."""
    board, _, _ = _get_board()
    filters: dict = {}
    if sprint:
        filters["sprint"] = sprint
    if status:
        filters["status"] = status
    if agent:
        filters["agent"] = agent
    if tag:
        filters["tag"] = tag
    if scope:
        filters["scope"] = scope
    if priority and priority.startswith("p") and len(priority) == 2:
        filters["priority"] = priority

    tickets = board.ls(filters)
    if not tickets:
        console.print("[dim]No tickets match the filters.[/dim]")
        return

    for t in tickets:
        dep = f" [dim][dep: {', '.join(t['depends'])}][/dim]" if t.get("depends") else ""
        agents = ", ".join(t["agent"]) if t.get("agent") else "—"
        agents_str = f"[green]{agents}[/green]"
        console.print(
            f"[bold]{t['id']}[/bold]  {_priority_color(t['priority'])}  "
            f"{_status_color(t['status'])}  {(t.get('sprint') or '—'):<12}  "
            f"{agents_str:<20}  {t['title'][:50]}{dep}"
        )


@app.command("move")
def move_cmd(
    ticket_id: str = typer.Argument(...),
    status: str = typer.Argument(...),
):
    """Move a ticket to a new status."""
    board, _, _ = _get_board()
    try:
        result = board.move(ticket_id, status)
        console.print(f"{result['id']}: {result['from']} → [bold]{result['to']}[/bold]")
    except (ValueError, KeyError) as e:
        console.print(f"[red]{e}[/red]")
        raise typer.Exit(1)


@app.command("set")
def set_cmd(
    ticket_id: str = typer.Argument(...),
    field: str = typer.Argument(...),
    value: list[str] = typer.Argument(...),
):
    """Set a field on a ticket."""
    board, _, _ = _get_board()
    try:
        result = board.set(ticket_id, field, " ".join(value))
        console.print(f"{result['id']}.{result['field']} = {json.dumps(result['value'])}")
    except KeyError as e:
        console.print(f"[red]{e}[/red]")
        raise typer.Exit(1)


@app.command("add")
def add_cmd(ticket_json: str = typer.Argument(..., help="JSON ticket data")):
    """Create a new ticket from JSON."""
    board, _, _ = _get_board()
    try:
        patch = json.loads(ticket_json)
        ticket = board.add(patch)
        console.print(f"[green]Created {ticket['id']}: {ticket['title']}[/green]")
        print(json.dumps(ticket, indent=2))
    except (json.JSONDecodeError, Exception) as e:
        console.print(f"[red]{e}[/red]")
        raise typer.Exit(1)


@app.command("next-id")
def next_id_cmd():
    """Show the next available ticket ID."""
    board, _, _ = _get_board()
    print(board.next_id())


@app.command("rebuild-index")
def rebuild_index_cmd():
    """Rebuild index.json from ticket .md files."""
    board, _, _ = _get_board()
    result = board.rebuild_index()
    console.print(f"[green]Rebuilt index: {result['ticketCount']} tickets, nextId: {result['nextId']}[/green]")


def _priority_color(p: str) -> str:
    colors = {"p0": "red", "p1": "yellow", "p2": "blue", "p3": "dim"}
    color = colors.get(p, "white")
    return f"[{color}]{p:<2}[/{color}]"


def _status_color(s: str) -> str:
    colors = {"backlog": "dim", "doing": "cyan", "review": "yellow", "done": "green", "cancelled": "strike"}
    color = colors.get(s, "white")
    return f"[{color}]{s:<10}[/{color}]"
